js/ui/grid_core/ui.grid_core.editing.js

```diff
@@ -1655,130 +1680,151 @@ const EditingController = modules.ViewController.inherit((function() {
+            when(this._afterSaveEditData(cancel)).done(function() {
+                deferred.resolve(error);
+            }).fail(deferred.reject);
+        },
+
         _saveEditDataInner: function() {
             const results = [];
             const deferreds = [];
             const changes = [];
             const dataController = this._dataController;
             const dataSource = dataController.dataSource();
-            const editMode = getEditMode(this);
             const result = new Deferred();
             const editData = this._editData.slice(0);
 
-            const resetEditIndices = () => {
-                if(editMode !== EDIT_MODE_CELL) {
-                    this._resetEditColumnName();
-                    this._resetEditRowKey();
-                }
-            };
-            const resetModifiedClassCells = () => {
-                if(editMode === EDIT_MODE_BATCH) {
-                    const columnsCount = this._columnsController.getVisibleColumns().length;
-                    editData.forEach(({ key }) => {
-                        const rowIndex = this._dataController.getRowIndexByKey(key);
-                        if(rowIndex !== -1) {
-                            for(let columnIndex = 0; columnIndex < columnsCount; columnIndex++) {
-                                this._rowsView._getCellElement(rowIndex, columnIndex).removeClass(CELL_MODIFIED);
-                            }
+            when(this._saveEditDataCore(deferreds, results, changes)).done(() => {
+                if(deferreds.length) {
+                    dataSource?.beginLoading();
+
+                    when(...deferreds).done(() => {
+                        if(this._processSaveEditDataResult(results)) {
+                            this._endSaving(changes, editData, result);
+                        } else {
+                            dataSource?.endLoading();
+                            result.resolve();
                         }
+                    }).fail(function(error) {
+                        dataSource?.endLoading();
+                        result.resolve(error);
                     });
-                }
-            };
-            const afterSaveEditData = (error) => {
-                when(this._afterSaveEditData()).done(function() {
-                    result.resolve(error);
-                });
-            };
 
-            if(!this._saveEditDataCore(deferreds, results, changes) && editMode === EDIT_MODE_CELL) {
-                this._focusEditingCell();
-            }
+                    this.executeAction('onSaved');
+                    return result.always(() => {
+                        this._focusEditingCell();
+                        this._saving = false;
+                    }).promise();
+                }
 
-            if(deferreds.length) {
-                this._saving = true;
+                this._cancelSaving();
+                this._resolveAfterSaveEditDataComplete(result);
```

Comment:
I think that _resolveAfterSaveEditDataComplete should be called inside _cancelSaving similar to _endSaving.